WooCommerce/Classes/ViewRelated/Blaze/BudgetSetting/BlazeBudgetSettingViewModel.swift

```diff
@@ -39,33 +38,110 @@ final class BlazeBudgetSettingViewModel: ObservableObject {
         formatter.timeStyle = .none
         return formatter
     }()
 
-    var formattedDateRange: String {
-        let endDate = Date(timeInterval: Constants.oneDayInSeconds * dayCount, since: startDate)
+    private let siteID: Int64
+    private let targetOptions: BlazeTargetOptions?
+    private let stores: StoresManager
 
-        // Use the configured formatter to generate the string.
-        return dateFormatter.string(from: startDate, to: endDate)
-    }
+    typealias BlazeBudgetSettingCompletionHandler = (_ dailyBudget: Double, _ duration: Int, _ startDate: Date) -> Void
+    private let completionHandler: BlazeBudgetSettingCompletionHandler
+
+    private var settingSubscription: AnyCancellable?
 
-    init(dailyBudget: Double, duration: Int, startDate: Date, onCompletion: @escaping BlazeBudgetSettingCompletionHandler) {
+    init(siteID: Int64,
+         dailyBudget: Double,
+         duration: Int,
+         startDate: Date,
+         targetOptions: BlazeTargetOptions? = nil,
+         stores: StoresManager = ServiceLocator.stores,
+         onCompletion: @escaping BlazeBudgetSettingCompletionHandler) {
+        self.siteID = siteID
         self.dailyAmount = dailyBudget
         self.dayCount = Double(duration)
         self.startDate = startDate
+        self.targetOptions = targetOptions
+        self.stores = stores
         self.completionHandler = onCompletion
+
+        observeSettings()
     }
 
     func confirmSettings() {
         // TODO: track confirmation
         completionHandler(dailyAmount, Int(dayCount), startDate)
     }
+
+    @MainActor
+    func retryFetchingImpressions() async {
+        await updateImpressions(startDate: startDate, dayCount: dayCount, dailyBudget: dailyAmount)
+    }
+
+    @MainActor
+    func updateImpressions(startDate: Date, dayCount: Double, dailyBudget: Double) async {
+        forecastedImpressionState = .loading
+        let endDate = calculateEndDate(from: startDate, dayCount: dayCount)
+        let totalBudget = calculateTotalBudget(dailyBudget: dailyBudget, dayCount: dayCount)
+        let input = BlazeForecastedImpressionsInput(startDate: startDate, endDate: endDate, totalBudget: totalBudget, targetings: targetOptions)
+        do {
+            let result = try await fetchForecastedImpressions(input: input)
+            let formattedImpressions = String(format: "%d - %d", result.totalImpressionsMin, result.totalImpressionsMax)
+            forecastedImpressionState = .result(formattedResult: formattedImpressions)
+        } catch {
+            DDLogError("⛔️ Error fetching forecasted impression: \(error)")
+            forecastedImpressionState = .failure
+        }
+    }
 }
 
+// MARK: - Private helpers
+private extension BlazeBudgetSettingViewModel {
+
+    func observeSettings() {
+        settingSubscription = $dayCount.combineLatest($dailyAmount, $startDate)
+            .debounce(for: .milliseconds(300), scheduler: DispatchQueue.main)
```

Comment:
Nice, I was wondering how to best prevent calling impressions value too frequently given that a slider can be moved rather quickly.